src/utils.jl

```diff
@@ -107,3 +130,66 @@ end
   isnothing(ret) && return IteratorReturn(nothing)
   ret
 end
+
+# this is similar to code_typed but it considers the world age
+function code_typed_by_type(@nospecialize(tt::Type);
+                            optimize::Bool=true,
+                            world::UInt=Base.get_world_counter(),
+                            interp::Core.Compiler.AbstractInterpreter=Core.Compiler.NativeInterpreter(world))
+    tt = Base.to_tuple_type(tt)
+    # look up the method
+    match, valid_worlds = Core.Compiler.findsup(tt, Core.Compiler.InternalMethodTable(world))
+    meth = Base.func_for_method_checked(match.method, tt, match.sparams)
+    # run inference, normally not allowed in generated functions
+    frame = Core.Compiler.typeinf_frame(interp, meth, match.spec_types, match.sparams, optimize)
+    frame === nothing && error("inference failed")
+    return frame.linfo, frame.src, valid_worlds
+end
+
+function fsmi_generator(world::UInt, source::LineNumberNode, passtype, fsmitype::Type{Type{T}}, fargtypes) where T
+    @nospecialize
+    # get typed code of the inference function evaluated in get_slots
+    # but this time with concrete argument types
+    tt = Base.to_tuple_type(fargtypes)
+    mi, ci, valid_worlds = code_typed_by_type(tt; world, optimize=false)
+    (; min_world, max_world) = valid_worlds
+    # extract slot types
+    cislots = Dict(zip(ci.slotnames, ci.slottypes))
+    slots = map(fieldnames(T)[2:end]) do slot
+      s = get(cislots, slot, Any)
+      Core.Compiler.widenconst(s)
+    end
+    # generate code to instantiate the concrete type
+    stub = Core.GeneratedFunctionStub(identity, Core.svec(:pass, :fsmi, :fargs), Core.svec())
+    if isempty(slots)
+      exprs = stub(world, source, :(return $T()))
+    else
+      exprs = stub(world, source, :(return $T{$(slots...)}()))
+    end
+    # lower codeinfo to pass world age and invalidation edges
+    ci = ccall(:jl_expand_and_resolve, Any, (Any, Any, Any), exprs, passtype.name.module, Core.svec())
+    ci.min_world = min_world
+    ci.max_world = max_world
+    ci.edges = Core.MethodInstance[mi]
+    return ci
```

Comment:
```suggestion
        mi, ci, (; min_world, max_world) = try
      code_typed_by_type(tt; world, optimize=false)
    catch
      @goto fallback_impl
    end
    # extract slot types
    cislots = Dict(zip(ci.slotnames, ci.slottypes))
    slots = map(fieldnames(T)[2:end]) do slot
      s = get(cislots, slot, Any)
      Core.Compiler.widenconst(s)
    end
    # generate code to instantiate the concrete type
    stub = Core.GeneratedFunctionStub(identity, Core.svec(:pass, :fsmi, :fargs), Core.svec())
    if isempty(slots)
      @label fallback_impl
      exprs = stub(world, source, :(return $T()))
    else
      exprs = stub(world, source, :(return $T{$(slots...)}()))
    end
    # lower codeinfo to pass world age and invalidation edges
    ci = ccall(:jl_expand_and_resolve, Any, (Any, Any, Any), exprs, passtype.name.module, Core.svec())
    ci.min_world = min_world
    ci.max_world = max_world
    ci.edges = Core.MethodInstance[mi]
    return ci
```

Reply:
I don't think that's quite right since it'll try to instantiate an abstract type, but something along those lines yeah

Reply:
My struggle is what to do with he world age when inference fails.